woocommerce/woocommerce
# Remove the Legacy REST API from WooCommerce core

### Changes proposed in this Pull Request:

This pull request implements the removal of the Legacy REST API from WooCommerce core, [as announced in the WooCommerce developers blog](https://developer.woocommerce.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/).

#### Endpoint changes

Any request for any endpoint whose URL starts with `/wc-api/v3` will now return a 404 error with the following contents, exactly in the same way as when the Legacy REST API was disabled in WooCommerce 8.9 and older:

```json
{
    "errors": {
        "code": "woocommerce_api_disabled",
        "message": "The WooCommerce API is disabled on this site"
    }
}
```

For compatibility, the Legacy REST API will be considered disabled (endpoints won't work) if the `woocommerce_api_enabled` is not set to `yes`, even if the plugin is active.

#### Changes in the Legacy API settings page

The WooCommerce - Advanced - Legacy API settings page will now tell whether the plugin is installed or not, but it won't be possible to actually change anything from that page:

![image](https://github.com/woocommerce/woocommerce/assets/937723/18677614-cc3c-4e46-8b63-16ee263a20ed)

![image](https://github.com/woocommerce/woocommerce/assets/937723/264306de-1f9b-4c06-8b23-4d1b544e0ddd)

#### Changes in the webhooks page

If one or more webhooks exist that are configured for using the Legacy REST API for composing their payloads and the Legacy REST API plugin is **not** active, a warning notice is shown in the  WooCommerce - Advanced - Webhooks page and a warning icon with a tooltip prepends the name of the "offending" webhooks:

![image](https://github.com/woocommerce/woocommerce/assets/937723/71a3d3f8-6c26-4549-aa5b-047da89be3a6)

Also, the wording of the already existing global admin notice about these webhooks has changed from "...will be unsupported soon" to "...is unsupported".

![image](https://github.com/woocommerce/woocommerce/assets/937723/0fbc7882-e990-4c09-876f-1a14e0fcc6c4)

Additionally, a "NOT AVAILABLE" warning will be displayed in the format dropdown in the webhook edit page when the legacy API payload format is selected (if the Legacy API plugin is installed, the warning text is "deprecated"):

![image](https://github.com/woocommerce/woocommerce/assets/937723/91f29725-77de-480f-b354-9152bb27fbb6)

![image](https://github.com/woocommerce/woocommerce/assets/937723/4561274e-6aa9-43eb-90d4-f53bc23b2462)

#### Changes in the status report

The WooCommerce status report (WooCommerce - Status - System status) will display the version of the installed Legacy REST API plugin, or a message indicating that the plugin is not installed:

![image](https://github.com/woocommerce/woocommerce/assets/937723/19d7bbea-305c-41be-b07e-ebbd8e20e584)

![image](https://github.com/woocommerce/woocommerce/assets/937723/6e7e6104-dd3e-45ea-af85-368bb817fa3e)

### How to test the changes in this Pull Request:

Test upgrading from an older WooCommerce to a build containing this pull request, and verifying that the Legacy REST API works or not depending on all the sensible combinations (the first three columns refer to the WooCommerce state _before_ the upgrade, the last two are _after_ the upgrade):

Legacy REST API active | Legacy webhook exists | Plugin was already installed | Plugin gets autoinstalled | Legacy REST API works
|-|-|-|-|-|
❌ | ❌ | ❌ | ❌ | ❌
❌ | ❌ | ✔️ | ❌ | ✔️
✔️ | ❌ | ✔️ | ❌ | ✔️
❌ | ✔️ | ✔️ | ❌ | ✔️
✔️ | ✔️ | ✔️ | ❌ | ✔️
✔️ | ❌ | ❌ | ✔️ | ✔️
❌ | ✔️ | ❌ | ✔️ | ✔️
✔️ | ✔️ | ❌ | ✔️ | ✔️

Notes:

* If the plugin gets autoinstalled, you'll get an admin notice and a message next to the plugin name in the plugins page (see https://github.com/woocommerce/woocommerce/pull/45570)
* "Legacy REST API works" means that you get the expected result when querying any of the `/wc-api/v3/` entpoints (e.g. `/wc-api/v3/orders`); the opposite means that you get the "The WooCommerce API is disabled on this site" message.
* If you are switching back to an older version of WooCommerce as part of your testing, remember that the value of the `wc_version` option won't be downgraded: you need to do it manually (e.g. `wc option set 8.7.0`) or the WooCommerce upgrade code won't run the next time you upgrade again to a higher version!

There's an extra test to perform. The store API was using `wc()->api->get_endpoint_data` to get the current cart under certain error conditions when hitting the checkout endpoint, this has been replaced with `wc_get_container()->get( RestApiUtil::class )->get_endpoint_data` which is equivalent. The easiest way to test this is to perform a temporary modification in the `Automattic\WooCommerce\StoreApi\Routes\V1\Checkout` class, rewriting the `get_response` method as follows:

```php
public function get_response( \WP_REST_Request $request ) {
     $this->load_cart_session( $request );
     $this->cart_controller->calculate_totals();
     $error = new \WP_Error('an_error', 'This is an error');
     $this->add_data_to_error_object($error, ['some_data' => 'hello'], 400, true);
     $error = $this->error_to_response( $error );
     return $this->add_response_headers( $error );
}
```

With this change, run `curl http://localhost/wp-json/wc/store/v1/checkout` and verify that the response contains information for an empty cart.
